File: liegw/src/lielas/modul.c
/*
*
*	Copyright (c) 2013 Andreas Reder
*	Author      : Andreas Reder <andreas@reder.eu>
*	File		: 
*
*	This File is part of lieweb, a web-gui for osd-compatible devices
*
*   This program is free software: you can redistribute it and/or modify
*   it under the terms of the GNU General Public License as published by
*   the Free Software Foundation, either version 3 of the License, or
*   (at your option) any later version.
*
*   This program is distributed in the hope that it will be useful,
*   but WITHOUT ANY WARRANTY; without even the implied warranty of
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*   GNU General Public License for more details.
*
*   You should have received a copy of the GNU General Public License
*   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*
*/


#include "modul.h"
#include "../log.h"

#include <stdlib.h>
#include <stdio.h>

Lmodul *LcreateModul(){
	Lmodul *m = malloc(sizeof(Lmodul));
	if(m == NULL){
		lielas_log((unsigned char*)"Can't allocate memory for new modul", LOG_LEVEL_ERROR);
		return NULL;
	}
	m->id = 0;
	m->address[0] = 0;
	m->channels = 0;
	m->channel[0] = NULL;
	return m;
}

void LdeleteModul(Lmodul *m){
	int i;

	for( i = 0; i < m->channels; i++){
		LdeleteChannel(m->channel[i]);

	}
	free(m);
}

void LaddChannel(Lmodul *m, Lchannel *channel){
	if(m->channels == MAX_CHANNELS){
		lielas_log((unsigned char*)"Can't add channel, maximum number of channels reached", LOG_LEVEL_WARN);
		return;
	}
	m->channel[m->channels+1] = channel;
  m->channels += 1;
	if(m->channels <= MAX_CHANNELS){
		m->channel[m->channels+1] = NULL;
	}
}

void LremoveChannel(Lmodul *m, Lchannel *channel){
	Lchannel *c;
	int i;
	int cIndex;

	//search channel
	for( cIndex = 0; cIndex < MAX_CHANNELS && cIndex < m->channels; cIndex++){
		if(m->channel[cIndex] == channel){
			break;
		}
	}
	if(m->channel[cIndex] != channel){
		lielas_log((unsigned char*)"Can't remove channel, channel not found", LOG_LEVEL_WARN);
		return;
	}

	c = m->channel[cIndex];

	// order channels if not the last channel will be removed
	if(cIndex != (m->channels - 1)){
		for( i = cIndex; i < m->channels; i++){
			m->channel[i] = m->channel[i+1];
		}
	}

	//remove channel
	LdeleteChannel(c);
	m->channels -= 1;
}

File: lieweb/src/lielas/core/Modul.java
/*
*
*	Copyright (c) 2013 Andreas Reder
*	Author      : Andreas Reder <andreas.reder@lielas.org>
*	File		: 
*
*	This File is part of lielas, see www.lielas.org for more information.
*
*   This program is free software: you can redistribute it and/or modify
*   it under the terms of the GNU General Public License as published by
*   the Free Software Foundation, either version 3 of the License, or
*   (at your option) any later version.
*
*   This program is distributed in the hope that it will be useful,
*   but WITHOUT ANY WARRANTY; without even the implied warranty of
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*   GNU General Public License for more details.
*
*   You should have received a copy of the GNU General Public License
*   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*
*/
spackage lielas.core;

import java.io.Serializable;
import lielas.core.Channel;

@SuppressWarnings("serial")

public class Modul implements Serializable{

	private int address;
	private int meassurementIntervall;
	private int processIntervall;
	private int alarmIntervall;
	private Channel[] channel;
	private int id;

	private int channels;
	
	public Modul(int id, int mInt, int pInt, int aInt) {
		super();
		this.id = id;
		this.meassurementIntervall = mInt;
		this.processIntervall = pInt;
		this.alarmIntervall = aInt;
		this.channels = 0;
		
		channel =  new Channel[20];
	}

	public Integer getAddress() {
		return address;
	}

	public void setAddress(Integer address) {
		this.address = address;
	}

	public Integer getMeassurementIntervall() {
		return meassurementIntervall;
	}

	public void setMeassurementIntervall(Integer meassurementIntervall) {
		this.meassurementIntervall = meassurementIntervall;
	}

	public Integer getProcessIntervall() {
		return processIntervall;
	}

	public void setProcessIntervall(Integer processIntervall) {
		this.processIntervall = processIntervall;
	}

	public Integer getAlarmIntervall() {
		return alarmIntervall;
	}

	public void setAlarmIntervall(Integer alarmIntervall) {
		this.alarmIntervall = alarmIntervall;
	}

	public Integer getChannels() {
		return channels;
	}

	public void setChannels(Integer channels) {
		this.channels = channels;
	}
	
	public void addChannel(Channel c){
		channel[c.getAddress()] = new Channel(c.getID(), c.getAddress(), c.getType(), c.getUnit(), c.getMinValue(), c.getMaxValue());
		channel[c.getAddress()].setName(c.getName());
		channel[c.getAddress()].setGroup(c.getGroup());
		this.channels += 1;
	}
	
	public Channel getChannel(int address){
		return channel[address];
	}
	
	public int getID(){
		return this.id;
	}
	
	static public String getProcessIntervallString(int pInt){
		switch(pInt){
		case 1:
			return LanguageHelper.AT_PINT_OPT_LIFETIME;
		case 2:
			return LanguageHelper.AT_PINT_OPT_NORMAL;
		case 3:
			return LanguageHelper.AT_PINT_OPT_FAST_GATHER;
		}
		return null;
	}
	
	static public int getProcessIntervallInt(String pInt, LanguageHelper langHelper){
		if(pInt.equals(langHelper.GetString(LanguageHelper.AT_PINT_OPT_LIFETIME))){
			return 1;
		}else if(pInt.equals(langHelper.GetString(LanguageHelper.AT_PINT_OPT_NORMAL))){
			return 2;
		}else if(pInt.equals(langHelper.GetString(LanguageHelper.AT_PINT_OPT_FAST_GATHER))){
			return 3;
		}
		return 0;
	}
	
}
